polkadot/consensus/src/shared_table/mod.rs

```diff
@@ -0,0 +1,566 @@
+			&& self.checked_validity.insert(digest.clone());
+
+		let checking_availability = is_availability_member
+			&& self.checked_availability.insert(digest.clone());
+
+		let work = if checking_validity || checking_availability {
+			match self.table.get_candidate(&digest) {
+				None => None, // TODO: handle table inconsistency somehow?
+				Some(candidate) => {
+					let fetch_block_data =
+						router.fetch_block_data(candidate).into_future().fuse();
+
+					let fetch_extrinsic = if checking_availability {
+						Some(
+							router.fetch_extrinsic_data(candidate).into_future().fuse()
+						)
+					} else {
+						None
+					};
+
+					Some(Work {
+						candidate_receipt: candidate.clone(),
+						fetch_block_data,
+						fetch_extrinsic,
+						evaluate: checking_validity,
+						check_candidate,
+					})
+				}
+			}
+		} else {
+			None
+		};
+
+		StatementProducer {
+			produced_statements: Default::default(),
+			work,
+		}
+	}
+
+	fn update_trackers(&mut self, candidate: &Hash, context: &TableContext) {
+		let includable = self.table.candidate_includable(candidate, context);
+		for i in (0..self.trackers.len()).rev() {
+			if self.trackers[i].update_candidate(candidate.clone(), includable) {
+				self.trackers.swap_remove(i);
+			}
+		}
+	}
+}
+
+/// Produced statements about a specific candidate.
+/// Both may be `None`.
+#[derive(Default)]
+pub struct ProducedStatements {
+	/// A statement about the validity of the candidate.
+	pub validity: Option<table::Statement>,
+	/// A statement about availability of data. If this is `Some`,
+	/// then `block_data` and `extrinsic` should be `Some` as well.
+	pub availability: Option<table::Statement>,
+	/// Block data to ensure availability of.
+	pub block_data: Option<BlockData>,
+	/// Extrinsic data to ensure availability of.
+	pub extrinsic: Option<Extrinsic>,
+}
+
+/// Future that produces statements about a specific candidate.
+pub struct StatementProducer<D: Future, E: Future, C> {
+	produced_statements: ProducedStatements,
+	work: Option<Work<D, E, C>>,
+}
+
+struct Work<D: Future, E: Future, C> {
+	candidate_receipt: CandidateReceipt,
+	fetch_block_data: future::Fuse<D>,
+	fetch_extrinsic: Option<future::Fuse<E>>,
+	evaluate: bool,
+	check_candidate: C
+}
+
+impl<D: Future, E: Future, C> Default for StatementProducer<D, E, C> {
+	fn default() -> Self {
+		StatementProducer {
+			produced_statements: Default::default(),
+			work: None,
+		}
+	}
+}
+
+impl<D, E, C, Err> Future for StatementProducer<D, E, C>
+	where
+		D: Future<Item=BlockData,Error=Err>,
+		E: Future<Item=Extrinsic,Error=Err>,
+		C: FnMut(Collation) -> bool,
+{
+	type Item = ProducedStatements;
+	type Error = Err;
+
+	fn poll(&mut self) -> Poll<ProducedStatements, Err> {
+		let work = match self.work {
+			Some(ref mut work) => work,
+			None => return Ok(Async::Ready(::std::mem::replace(&mut self.produced_statements, Default::default()))),
+		};
+
+		if let Async::Ready(block_data) = work.fetch_block_data.poll()? {
```

Comment:
We do we ignore `NotReady` here an couple of lines below? Woulnd't that break if that other futures are made truly async?

Reply:
Oh ok, now I see how it works - we first poll both and if they don't supply data we assume `NotReady` anyway. So that if the futures are fused we can poll them multiple times.

Wouldn't it be easier to have a classical `and_then` chaining done inside `StatementProducer::new/default`?

Reply:
I don't think so, at least not without using `Box<Future>`. I would expect this to be a fairly hot path so I would prefer to avoid that.